data_structures/treap.cpp

```diff
@@ -0,0 +1,223 @@
+     * @param x Query the subtree of node x (Usually x=root)
+     * @param k The queried rank
+     * @return The element ranked number k
+     */
+    int _get_k_th(int &x, int k) {
+        if (k <= size[childs[x][0]]) {
+            return _get_k_th(childs[x][0], k);
+        }
+        k -= size[childs[x][0]] + cnt[x];
+        if (k <= 0) {
+            return key[x];
+        }
+        return _get_k_th(childs[x][1], k);
+    }
+    /**
+     * @brief Query the rank of specified element (internal method)
+     * @param x Query the subtree of node x (Usually x=root)
+     * @param k The queried element
+     * @return The rank of element k
+     */
+    int _get_rank(int x, int k) {
+        if (!x)
+            return 0;
+        if (k == key[x])
+            return size[childs[x][0]] + 1;
+        else if (k < key[x])
+            return _get_rank(childs[x][0], k);
+        else
+            return size[childs[x][0]] + cnt[x] + _get_rank(childs[x][1], k);
+    }
+    /**
+     * @brief Get the predecessor node of element k
+     * @param k The queried element
+     * @return The predecessor
+     */
+    int get_predecessor(int k) {
+        int x = root, pre;
+        while (x) {
+            if (key[x] < k) {
+                pre = key[x], x = childs[x][1];
+            } else
+                x = childs[x][0];
+        }
+        return pre;
+    }
+    /**
+     * @brief Get the successor node of element k
+     * @param k The queried element
+     * @return The successor
+     */
+    int get_next(int k) {
+        int x = root, next;
+        while (x) {
+            if (key[x] > k) {
+                next = key[x], x = childs[x][0];
+            } else
+                x = childs[x][1];
+        }
+        return next;
+    }
+    /**
+     * @brief Insert element (External method)
+     * @param k Key to insert
+     */
+    void insert(int k) { _insert(root, k); }
+    /**
+     * @brief Erase element (External method)
+     * @param k Key to erase
+     */
+    void erase(int k) { _erase(root, k); }
+    /**
+     * @brief Get the KTH largest value (External method)
+     * @param k The queried rank
+     * @return The element ranked number x
+     */
+    int get_k_th(int k) { return _get_k_th(root, k); }
+    /**
+     * @brief Get the rank of specified element (External method)
+     * @param k The queried element
+     * @return The rank of element k
+     */
+    int get_rank(int k) { return _get_rank(root, k); }
+} mTreap;
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    std::ios::sync_with_stdio(0);
+    std::cin.tie(0);
+    std::cout.tie(0);
+
+    mTreap.insert(106465);
+    assert(mTreap.get_k_th(1) == 106465);
+    mTreap.insert(317721);
+    mTreap.insert(460929);
+    mTreap.insert(644985);
+    mTreap.insert(84185);
+    mTreap.insert(89851);
+    assert(mTreap.get_next(81968) == 84185);
+    mTreap.insert(492737);
+    assert(mTreap.get_predecessor(493598) == 492737);
+}
```

Comment:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
 static void test() 
    mTreap.insert(106465);
    assert(mTreap.get_k_th(1) == 106465);
    mTreap.insert(317721);
    mTreap.insert(460929);
    mTreap.insert(644985);
    mTreap.insert(84185);
    mTreap.insert(89851);
    assert(mTreap.get_next(81968) == 84185);
    mTreap.insert(492737);
    assert(mTreap.get_predecessor(493598) == 492737);{
 }
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
    std::ios::sync_with_stdio(0);
    std::cin.tie(0);
    std::cout.tie(0);
    test(); // run self-test implementations
    return 0;
}
```
tests must always be separate from the main function